packages/compile-solidity/src/run.ts

```diff
@@ -1,14 +1,81 @@
-const debug = require("debug")("compile:run");
-const OS = require("os");
-const semver = require("semver");
-const Common = require("@truffle/compile-common");
-const { CompilerSupplier } = require("./compilerSupplier");
-const { zeroLinkReferences, formatLinkReferences } = require("./shims");
+import { zeroLinkReferences, formatLinkReferences } from "./shims";
+import debugModule from "debug";
+const debug = debugModule("compile:run");
+import OS = require("os");
+import semver from "semver";
+import { CompilerSupplier } from "./compilerSupplier";
+import type {
+  Compilation,
+  Source,
+  CompiledContract
+} from "@truffle/compile-common";
+import type Config from "@truffle/config";
+import * as Common from "@truffle/compile-common";
+
+type Targets = string[];
+
+type CompilerOutput = {
+  contracts: {
+    [path: string]: object;
+  };
+  sources: {
+    [path: string]: {
+      ast?: object;
+      legacyAST?: object;
+      id: number;
+    };
+  };
+  errors?: any[];
+};
+
+type ProcessAllSourcesArgs = {
+  sources: Common.Sources.Sources;
+  compilerOutput: CompilerOutput;
+  originalSourcePaths: any;
+  language: string;
+};
+
+type PrepareCompilerInputArgs = {
+  sources: Common.Sources.Sources;
+  targets: Targets;
+  language: string;
+  settings: any;
+  modelCheckerSettings: any;
+};
+
+type InternalOptions = {
+  language?: string;
+  noTransform?: boolean;
+  solc?: any;
+};
+
+type PreparedSources = {
+  [path: string]: {
+    content: string;
+  };
+};
+
+type ProcessContractsArgs = {
+  compilerOutput: CompilerOutput;
+  sources: Common.Sources.Sources;
+  originalSourcePaths: Common.Sources.PathMapping;
+  solcVersion: string;
+};
+
+type Contracts = {
+  [path: string]: {
+    [name: string]: any;
+  };
+};
 
 // this function returns a Compilation - legacy/index.js and ./index.js
 // both check to make sure rawSources exist before calling this method
 // however, there is a check here that returns null if no sources exist
-async function run(rawSources, options, internalOptions = {}) {
+export async function run(
+  rawSources: { [name: string]: string },
+  options: Config,
+  internalOptions: InternalOptions = {}
+): Promise<Compilation | null> {
   if (Object.keys(rawSources).length === 0) {
     return null;
   }
```

Comment:
If we are okay with breaking changes here, you _might_ want to consider making this throw instead of returning `null`, depending on what else calls this. That would be my recommendation, unless we do want this to be a strictly homomorphic change.

Reply:
Not important to me. I wanted to be sure that you'd considered it, but you _definitely_ have now, and it does seem reasonable to leave it as is.